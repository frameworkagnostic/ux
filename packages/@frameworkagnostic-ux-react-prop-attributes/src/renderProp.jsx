
import React from 'react';
import PropTypes from 'prop-types';

const PropAttributes = (props) => {
  const {
    attrs,
    children,
    render,
    renderComponent,
    Component,
    Components,
    inlineStyles,
    classNames,
    ...remainingProps
  } = props;

  if (renderComponent) {
    return React.createElement(renderComponent, {
      ...remainingProps,
      Component,
      Components,
      inlineStyles,
      classNames,
      attrs: attrs || remainingProps,
    }, children);
  }

  return render({
    ...remainingProps,
    children,
    Component,
    Components,
    inlineStyles,
    classNames,
    attrs: attrs || remainingProps,
  });
};

PropAttributes.propTypes = {
  attrs: PropTypes.objectOf(PropTypes.object),
  style: PropTypes.objectOf(PropTypes.string),
  children: PropTypes.node,
  inlineStyles: PropTypes.objectOf(PropTypes.object),
  className: PropTypes.string,
  classNames: PropTypes.objectOf(PropTypes.object),
  Component: PropTypes.bool,
  Components: PropTypes.bool,
  render: PropTypes.func,
  renderComponent: PropTypes.func,
};

PropAttributes.defaultProps = {
  attrs: null,
  children: null,
  style: {},
  inlineStyles: {},
  className: '',
  classNames: {},
  Component: undefined,
  Components: undefined,
  renderComponent: () => (
    <div>No render component provided</div>
  ),
  render: () => (
    <div>No render prop provided</div>
  ),
};

export { PropAttributes };
