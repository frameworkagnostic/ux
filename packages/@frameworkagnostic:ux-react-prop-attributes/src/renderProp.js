
import React from 'react';
import { appendPropTypes } from './utils/propTypeMeta';
import PropTypes from 'prop-types';

const PropAttributes = (props) => {
  const {
    attrs,
    children,
    renderProp,
    ...remaining,
  } = props;

  return renderProp(attrs ? {
    attrs,
    style: remaining.style,
    inlineStyles: remaining.inlineStyles,
    className: remaining.className,
    classNames: remaining.classNames,
    Component: remaining.Component,
    Components: remaining.Components,
  } : {
    attrs: remaining
  });
};

PropAttributes.propTypes = {
  children: PropTypes.oneOf([
    PropTypes.node,
    PropTypes.element,
  ]).isRequired,
  attrs: PropTypes.objectOf(PropTypes.object),
  style: PropTypes.objectOf(PropTypes.string),
  inlineStyles: PropTypes.objectOf(PropTypes.object),
  className: PropTypes.string,
  classNames: PropTypes.objectOf(PropTypes.object),
  Component: PropTypes.oneOf([
    PropTypes.bool,
    PropTypes.node,
    PropTypes.element,
  ]),
  Components: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.node,
    PropTypes.element,
  ]),
  renderProp: PropTypes.func,
};

PropAttributes.defaultProps = {
  attrs: null,
  style: {},
  inlineStyles: {},
  className: '',
  classNames: {},
  Component: null,
  Components: null,
  renderProp: () => (
    <div>No renderProp provided</div>
  ),
};

export { PropAttributes };
