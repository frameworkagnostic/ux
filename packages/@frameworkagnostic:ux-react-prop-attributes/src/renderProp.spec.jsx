import React from 'react';
import renderer from 'react-test-renderer';

import { PropAttributes } from './renderProp';

describe('<PropAttributes />', () => {
  test('validate element', () => expect(typeof PropAttributes).toBe('function'));

  describe('with no attributes', () => {
    const element = renderer.create(<PropAttributes />);
    const tree = element.toJSON();
    test('snapshot', () => expect(tree).toMatchSnapshot());
  });
});
