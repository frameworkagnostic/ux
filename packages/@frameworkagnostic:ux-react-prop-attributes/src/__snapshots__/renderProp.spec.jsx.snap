// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PropAttributes /> with no attributes snapshot 1`] = `
<div>
  No renderProp provided
</div>
`;
